Equivalence between LEM and DNE.

mr-infty opened the pull request:
It is a classical exercise in logic to show that the law of excluded
middle and double negation are equivalent.
Proving this formally in Idris makes it even more fun.

mr-infty committed:
Equivalence between LEM and DNE.

It is a classical exercise in logic to show that the law of excluded
middle and double negation are equivalent.
Proving this formally in Idris makes it even more fun.

stepancheg commented:
Both problems look very educating. It took me a dozen minutes to understand the statements.

Before I commit them can you please help explain it.

Why can't we prove "Double negation elimination" or "The law of the excluded middle" without the counterpart?

In other words, why can't we prove:

```
dne : DNE
dne = ?dne_rhs

lem : LEM
lem = ?lem_rhs
```

mr-infty commented:
If you could prove `LEM`, you would have found an algorithm that could decide every type, i.e. find an inhabitant or showing it‘s not inhabited. But such a thing doesn’t exist.

For example, you could easily translate the *mortal matrix problem*, which is known to be undecidable, into a type.

mr-infty committed:
Added Triple Negation Elimination.

Another classical exercise from logic which shows that even though
double negation elimination is unprovable in constructive logic, one can
nevertheless show that triple negations can be eliminated, in the sense
that they can be reduced to simple negations.

stepancheg commented:
> If you could prove LEM, you would have found an algorithm that could decide every type, i.e. find an inhabitant or showing it‘s not inhabited. But such a thing doesn’t exist.

Makes sense! Thank you for the explanation!

stepancheg commented:
Published.
